### PR TITLE
Workspace Resource Monitor Plugin

### PR DESCRIPTION
**Pull request by svor**
### What does this PR do?
This PR provides new built-in che-theia plugin **resource-monitor-plugin**.
The plugin shows information about used resources in workspace pod by using k8s API. For now it represents an information only about memory and CPU. 

_Note: Resource Monitor will be displayed only if Metric server was deployed and run on a cluster where Che is running. OpenShift 4 contains it by default and for Minishift and Minikube need to do it by hands._ 

The general information is displayed in Status Bar and it shows how much resources the workspace uses at all 
screenshot-nimbus-capture-2020 11 10-12_40_43
to see detailed information about each container need to click on that element in Status Bar
screenshot-nimbus-capture-2020 11 10-12_42_13

**Memory**
   For each container we can read Used value (from Metric server) and Limited (from workspace Pod description)
**CPU**
   Each container has Used value (from Metric server) but it is not possible to read Limited value if it wasn't set in workspace description. That's why status bar contains an information only about used CPU resources.

This PR is depends on https://github.com/eclipse/che-operator/pull/519 which customize cluster roles to make it possible to communicate with Metric server through k8s API. 

### Screenshot/screencast of this PR

ResMon
screenshot-che-che 192 168 99 180 nip io-2020 11 26-14_12_36
screenshot-che-che 192 168 99 182 nip io-2020 12 04-16_16_03
screenshot-che-che 172 17 0 2 nip io-2020 11 27-23_27_54

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/17205

### How to test this PR?

1. Make sure that Metric server is running in the cluster where Che will be deployed (OpenShift 4 provides it by default). 
For Minikube it could be done by:
 - `minikube addons enable metrics-server`
 - `kubectl -n kube-system rollout status deployment metrics-server`
2. Deploy Che by using **chectl** with these changes https://github.com/eclipse/che-operator/pull/519
3. Create a workspace by using this devfile:
<details>

<summary>Tested Devfile</summary>

```yaml
apiVersion: 1.0.0
metadata:
  name: golang-mrkd5
projects:
  - name: example
    source:
      location: 'https://github.com/golang/example.git'
      type: git
    clonePath: src/github.com/golang/example/
  - name: golang-echo-realworld-example-app
    source:
      location: 'https://github.com/xesina/golang-echo-realworld-example-app.git'
      type: git
    clonePath: src/github.com/xesina/golang-echo-realworld-example-app
components:
  - id: golang/go/latest
    memoryLimit: 512Mi
    preferences:
      go.lintFlags: '--fast'
      go.useLanguageServer: true
      go.lintTool: golangci-lint
    type: chePlugin
    alias: go-plugin
    env:
      - value: 'off'
        name: GO111MODULE
  - mountSources: true
    endpoints:
      - name: 8080-tcp
        port: 8080
    memoryLimit: 2Gi
    type: dockerimage
    image: 'quay.io/eclipse/che-golang-1.14:nightly'
    alias: go-cli
    env:
      - value: $(CHE_PROJECTS_ROOT)
        name: GOPATH
      - value: /tmp/.cache
        name: GOCACHE
  - reference: https://raw.githubusercontent.com/svor/che-plugin-registry/sv/testOperator/v3/plugins/eclipse/che-theia/next/meta.yaml
    type: cheEditor
commands:
  - name: 1.1 Run outyet
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/src/github.com/golang/example/outyet'
        type: exec
        command: go get -d && go run main.go
        component: go-cli
  - name: 1.2 Stop outyet
    actions:
      - type: exec
        command: kill $(pidof go)
        component: go-cli
  - name: 1.3 Test outyet
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/src/github.com/golang/example/outyet'
        type: exec
        command: go test
        component: go-cli
  - name: '2.1 xenisa :: install dependencies'
    actions:
      - workdir: '${GOPATH}/src/github.com/xesina/golang-echo-realworld-example-app'
        type: exec
        command: go mod download
        component: go-cli
  - name: '2.2 xenisa :: run'
    actions:
      - workdir: '${GOPATH}/src/github.com/xesina/golang-echo-realworld-example-app'
        type: exec
        command: go run main.go
        component: go-cli
  - name: '2.3 xenisa :: build'
    actions:
      - workdir: '${GOPATH}/src/github.com/xesina/golang-echo-realworld-example-app'
        type: exec
        command: go build
        component: go-cli
  - name: '2.4 xenisa :: test'
    actions:
      - workdir: '${GOPATH}/src/github.com/xesina/golang-echo-realworld-example-app'
        type: exec
        command: go test ./...
        component: go-cli
  - name: Run current file
    actions:
      - workdir: '${fileDirname}'
        type: exec
        command: 'go get -d && go run ${file}'
        component: go-cli
  - name: Debug current file
    actions:
      - referenceContent: |
          {
            "version": "0.2.0",
            "configurations": [
              {
                "name": "Debug current file",
                "type": "go",
                "request": "launch",
                "mode": "auto",
                "program": "${fileDirname}"
              }
            ]
          }
        type: vscode-launch
```

</details>

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=stable

### CODE REVIEW

**Comment by benoitf** [comment]
hello, would it make sense to expose these metrics from the `@eclipse/che-plugin` namespace ?
because other components might be interested in knowing the memory/cpu information ?

Main motivation is more about the k8s API access. It should be provided by che-theia, else each plug-in/extension will add 25Mi+ of dependencies. 
With dev workspace, k8s api will be a requirement as well, so in any case it will end it up into che-theia extensions, so it will ease each extension/plug-in to use that API.

**Comment by svor** [comment]
@benoitf Do you think k8s API will be needed for another plugins/extensions? Also [kubernetes-client/javascript](https://github.com/kubernetes-client/javascript) doesn't implement Metrics API yet, so this plugin uses kubernetes-client lib just to get cluster configuration and send raw queries to Kubernetes API.

**Comment by benoitf** [comment]
@svor yes basically with dev-workspace we'll do tons of stuff with k8s-api as all information is there
so all workspace-client stuff will goes through k8s. So: plug-in ext extension will require it.
I think recommendation plug-in might need as well the total memory available per pod (with k8s) and then there is workspace resource plug-in

I'm fine if for now we don't provide metrics but IMHO query api should be exposed as I see many upcoming requests.

**Comment by vitaliy-guliy** [comment]
I'm thinking about info in status bar. I think it's better to display the status only for one container? Displaying it overall for all containers is a bit confusing.
Giving a look at the status bar, I may think that I have 8GB memory, which will be anougn for everything. But it could happen that some process/plugin in some container will fail, because this container has a limitation in 256 mb.
Then the user will ask: why does it fail if I have 6GB memory?

File: extensions/eclipse-che-theia-plugin-ext/package.json [hunk 1]
```diff
@@ -19,6 +19,7 @@
     "@theia/mini-browser": "next",
     "@theia/plugin-ext": "next",
     "axios": "0.19.0",
+    "@kubernetes/client-node": "^0.12.1",
```
Comment (benoitf) [inline extensions/eclipse-che-theia-plugin-ext/package.json:22]: I think that I would expose kubernetes-api from a separate extension (like theia-plugin-remote extension that is exposing che-workspace-client to all extensions) and then re-expose it from plug-in-ext (so it's available from extension world and plug-ins world. You will argue that with inversify we can inject from anywhere but it's strange to have some extensions that would need to depends on plug-in-ext just to do k8s call)

I already extracted workspace-client from plug-in ext to cut the direct dependency so here we kinda doing again this dependency

TL;DR; I would move che-k8s-service.ts interface to a separate extension https://github.com/eclipse/che-theia/tree/master/extensions/eclipse-che-theia-remote-api/src/common and allow both client and server access

**Comment by benoitf** [comment]
I agree with @vitaliy-guliy that displaying the current/total memory is misleading as you might thing everything is OK while you have 99% of the memory of a container being used.

But then I don't know how to display something valuable with numbers like `current/total` Maybe a color indicator with a label changing from `Memory OK` in green to `Memory alert` and some limits being pre-defined like if one of current memory of a container is reaching more than 80% then the indicator becomes orange, then if one > 90% it becomes red, I don't know. 

In the detailed view, can we bring colors as well ? and be able to sort by 'most available memory' or 'consuming the most of cpu`

Surely UX team could help there

cc @beaumorley @parvathyvr

**Comment by vitaliy-guliy** [comment]
I would display detailed info somewhere in `MY WORKSPACE` panel.

**Comment by svor** [comment]
I decided to display general information about the workspace in total, because:
1/ If we show information only about some container, it is not clear which container it should be
2/ I don't want to add more text into status bar to describe the name of container, it will overload the status bar.

But I'm OK to display only some message like Memory OK/Memory alert with different colors in the status bar. Or just change the color of the text which i have now when some container will have little memory. 

About detailed info, as i know we can't bring different colors into QuickPick window which I use to represent it and also the content in the window doesn't update in real time.

File: extensions/eclipse-che-theia-plugin-ext/package.json (hunk 1 above)
Comment (svor) [reply extensions/eclipse-che-theia-plugin-ext/package.json]: updated

**Comment by svor** [comment]
About the info in the status bar. 
If some of workspace containers will use more than 90% of limited memory, then the information about this container will be displayed in status bar with warning color and the name of this container will be in the tooltip.
screenshot-che-che 192 168 99 180 nip io-2020 11 26-14_12_36
By default it will be the information about a workspace resources in total with tooltip **Workspace resources** 
screenshot-che-che 192 168 99 180 nip io-2020 11 26-14_04_24

**Comment by benoitf** [comment]
Hello, what is determining the unit used (like MB vs GB) ?
because it seems that the second screenshot with GB is taking less spaces (less digits)

`0.21/2.58` takes less space than `1083.23/1153.43` (I'm not sure to have this kind of precision is useful) (`1.0/1.1GB` might be enough or we drop as well the `.23` and `.43` when it's in MB format)

**Comment by svor** [comment]
@dmytro-ndp Hello! From happy-path-tests logs I see an error:
```log
15:33:47 error openid-client@4.2.1: The engine "node" is incompatible with this module. Expected version "^10.19.0 || >=12.0.0 < 13 || >=13.7.0 < 14 || >= 14.2.0". Got "10.15.3"
```
Looks like the problem with node version. Since we don't use node 12 for happy path job yet (https://github.com/eclipse/che/issues/18327), could we try to use node ^10.19.0 instead of 10.15.3. I think it could be a quicker solution than wait for https://github.com/eclipse/che/issues/18327?

**Comment by svor** [comment]
@benoitf Hello!
For now there is no special rules for determining units - GB for whole workspace, MB for containers. 
I'm going to add some logic for that (show in GB if it is > 1GB, otherwise show in MB and no extra precision for MB)

**Comment by svor** [comment]
Now if some container uses more than 90% of limited memory,  the message in status bar looks like:
screenshot-che-che 192 168 99 180 nip io-2020 11 26-20_58_26

**Comment by benoitf** [comment]
[crw-ci-test]

**Comment by benoitf** [comment]
I found a nodejs 12.18.2 install on the job and I applied it, trying...

**Comment by benoitf** [comment]
[crw-ci-test --rebuild]

**Comment by benoitf** [comment]
@svor trying the link on che.openshift.io I see nothing in the status bar ?

**Comment by benoitf** [comment]
[crw-ci-test]

**Comment by svor** [comment]
> @svor trying the link on che.openshift.io I see nothing in the status bar ?

@benoitf I think it is because of che was deployed without these changes:
https://github.com/che-incubator/chectl/pull/1002

**Comment by benoitf** [comment]
@svor: but we don't have any warning or like for eslint when it's not enabled (like a forbidden icon and if we click on it it's explains why it's not enabled) ?

**Comment by azatsarynnyy** [comment]
> @benoitf I think it is because of che was deployed without these changes:
> [eclipse/che-operator#519](https://github.com/eclipse/che-operator/pull/519)

I guess `HAPPY_PATH_CHANNEL=next` may help.

**Comment by svor** [comment]
> @svor: but we don't have any warning or like for eslint when it's not enabled (like a forbidden icon and if we click on it it's explains why it's not enabled) ?

Hmm, i think it makes sense to show some warning notification for these 2 cases (if some of them is true, resource monitor won't be displayed) :
- not enough roles to get workspace information (it can be fixed by updating che-operator)
- Metrics Server  is not enable in the cluster

**Comment by benoitf** [comment]
>> @benoitf I think it is because of che was deployed without these changes:
>> eclipse/che-operator#519

> I guess HAPPY_PATH_CHANNEL=next may help.

it won't as che.openshift.io is not redeployed on each PR. It's using a previous stable version of che.
it would help only if they were some selenium tests in Happy Path tests (and here we redeploy a new che cluster each time)

**Comment by svor** [comment]
in case if che doesn't have enough roles to get pod information or Metrics Server is not allowed, user will see ban icon on the status bar with tooltip Resources Monitor and if to click on that icon, a warning notification will be displayed:   
screenshot-codeready-che apps ci-ln-rhyh1nb-f76d1 origin-ci-int-gce dev openshift com-2020 11 30-14_08_44
screenshot-che-che 172 17 0 2 nip io-2020 11 27-23_27_54

**Comment by benoitf** [comment]
@svor great ! should it include 'Monitoring' label before or after the ban icon ?

**Comment by svor** [comment]
> @svor great ! should it include 'Monitoring' label before or after the ban icon ?

@benoitf it has a tooltip **Resourses Monitor**, do you think it should be a label with icon as well?

**Comment by benoitf** [comment]
eslint has always a label but it doesn't mean it's the right UX so probably will require feedback from UX team

**Comment by benoitf** [comment]
@svor so it seems adding a text around the ban icon would help

**Comment by svor** [comment]
@benoitf @vitaliy-guliy what do you think about this, is it better?
screenshot-che-che 192 168 99 182 nip io-2020 12 04-16_16_03

**Comment by benoitf** [comment]
cc @parvathyvr is it better ?

**Comment by parvathyvr** [comment]
> cc @parvathyvr is it better ?

@benoitf yes, I think having the 'Resources' near the ban icon makes it clear what the icon is indicating! cc @svor

**Review by ericwill — APPROVED** [review]
Anything else needed here? Or can we merge now.

**Comment by benoitf** [comment]
probably merge after 7.23

**Comment by ericwill** [comment]
> probably merge after 7.23

Any reason to wait? I'd like to show this at sprint review tomorrow.

**Comment by benoitf** [comment]
you can still show it at sprint demo. 
I think this kind of new plugin is better to be introduced at the beginning of the cycle than one day before the tag ( especially because it requires other dependencies in many other components and that we don't know if it'll work on che.openshift.io)
Also once merged anyone using Che can toggle the editor to next and see it live ( no need to wait extra 3 weeks)

**Comment by benoitf** [comment]
hey, I thought this PR was merged after 7.23 
Is it gonna happen ? There were tons of feedback when it was demoed but I think it can be addressed by iterating as we've few weeks without downstream releases

**Comment by svor** [comment]
Tried to run these changes on che.openshift.io and for some reason Resources monitor is  not displayed, i'm going to investigate why

**Comment by svor** [comment]
looks like there is a problem to read metrics for non-admin user:
```
kubectl get --raw /apis/metrics.k8s.io/v1beta1/namespaces/vsvydenk-che/pods/workspace0v3e2hcqhc8f8nvx.go-cli-66c57fcbb6-kwwjz
Error from server (Forbidden): pods.metrics.k8s.io "workspace0v3e2hcqhc8f8nvx.go-cli-66c57fcbb6-kwwjz" is forbidden: User "vsvydenk" cannot get pods.metrics.k8s.io in the namespace "vsvydenk-che": no RBAC policy matched
```

**Comment by svor** [comment]
@ibuziuk do you have some idea how we can allow non-admin users to read k8s metrics to avoid massage like:
```
{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"pods.metrics.k8s.io "workspace5elv4ll9wy7k47zw.go-cli-9844ddbf9-dfqlw" is forbidden: User "system:serviceaccount:vsvydenk-che:che-workspace" cannot get pods.metrics.k8s.io in the namespace "vsvydenk-che": no RBAC policy matched","reason":"Forbidden","details":{"name":"workspace5elv4ll9wy7k47zw.go-cli-9844ddbf9-dfqlw","group":"metrics.k8s.io","kind":"pods"},"code":403} 
```

**Comment by svor** [comment]
After testing this plugin on dogfooding instance I got a problem with an access to read pod information for service account:
screenshot-che-dogfooding apps che-dev x6e0 p1 openshiftapps com-2021 01 14-15_06_19

```
code: 403. Error: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"pods \"workspacegdve9us4ad860vrl.go-cli-85c4d6d888-mb7bc\" is forbidden: User \"system:serviceaccount:vsvydenk-che:che-workspace\" cannot get resource \"pods\" in API group \"\" in the namespace \"vsvydenk-che\"","reason":"Forbidden","details":{"name":"workspacegdve9us4ad860vrl.go-cli-85c4d6d888-mb7bc","kind":"pods"},"code":403}

    at ResMon.<anonymous> (/tmp/theia-unpacked/eclipse_che_resource_monitor_plugin.theia/lib/resource-monitor-plugin.js:119:35)
    at step (/tmp/theia-unpacked/eclipse_che_resource_monitor_plugin.theia/lib/resource-monitor-plugin.js:42:23)
    at Object.next (/tmp/theia-unpacked/eclipse_che_resource_monitor_plugin.theia/lib/resource-monitor-plugin.js:23:53)
    at fulfilled (/tmp/theia-unpacked/eclipse_che_resource_monitor_plugin.theia/lib/resource-monitor-plugin.js:14:58)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

probably service account needs to have more permission because it is possible to get Pod and Metrics information from the console where I logged in as vsvydenk user.

@ericwill @benoitf WDYT should I continue investigate this problem with this PR or it is better to merge it and open another issue?

**Comment by benoitf** [comment]
maybe if we don't have access we don't enable/display the plugin
so we can merge it quickly and user don't see errors on our hosted instances.
And we investigate in parallel

**Comment by ericwill** [comment]
> maybe if we don't have access we don't enable/display the plugin

Wouldn't it be better to load the plugin, but show the error message? Maybe the user wants to know about it.

> so we can merge it quickly and user don't see errors on our hosted instances.
> And we investigate in parallel

I am fine with merging now and iterating.

**Comment by svor** [comment]
agree with Eric to have a message that resources information is not available (which we can see on the screenshot above) shouldn't be a problem for the user. It is just a warning message if you click on the ban icon.
